lib/NoThrow.thy

```diff
@@ -0,0 +1,163 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ *)
+
+(* Lemmas about no_throw. Usually should have a conclusion "no_throw P m".
+   Includes some monad equations that have no_throw as a main assumption.  *)
+
+theory NoThrow
+  imports
+    WhileLoopRules
+    MonadEq_Lemmas
+begin
+
+(* Alternative definition of no_throw; easier to work with than unfolding validE. *)
+lemma no_throw_def': "no_throw P A = (\<forall>s. P s \<longrightarrow> (\<forall>(r, t) \<in> fst (A s). (\<exists>x. r = Inr x)))"
```

Comment:
what are your thoughts on having def' lemmas near their actual definitions so that people know that the def' versions exist?

Reply:
In general I'd agree, but here I would say that the definition of `no_throw` should eventually go into this file instead. 

If we do move it, it should be in a separate PR, I think.